src/core/lib/iomgr/timer_interface.h

```diff
@@ -0,0 +1,116 @@
+  // invoked. The application callback is also responsible for maintaining
+  // information about when to free up any user-level state
+  virtual void Init(grpc_exec_ctx* exec_ctx, TimerStorage* timer,
+                    grpc_millis deadline, grpc_closure* on_complete) = 0;
+
+  // Note that there is no timer destroy function. This is because the
+  // timer is a one-time occurrence with a guarantee that the callback will
+  // be called exactly once, either at expiration or cancellation. Thus, all
+  // the internal timer event management state is destroyed just before
+  // that callback is invoked. If the user has additional state associated with
+  // the timer, the user is responsible for determining when it is safe to
+  // destroy that state.
+
+  // Cancel a timer.
+  // There are three cases:
+  // 1. We normally cancel the timer
+  // 2. The timer has already run
+  // 3. We can't cancel the timer because it is "in flight".
+  //
+  // In all of these cases, the cancellation is still considered successful.
+  // They are essentially distinguished in that the timer_cb will be run
+  // exactly once from either the cancellation (with error GRPC_ERROR_CANCELLED)
+  // or from the activation (with error GRPC_ERROR_NONE).
+  //
+  // Note carefully that the callback function MAY occur in the same callstack
+  // as grpc_timer_cancel. It's expected that most timers will be cancelled
+  // (their primary use is to implement deadlines), and so this code is
+  // optimized such that cancellation costs as little as possible. Making
+  // callbacks run inline matches this aim.
+  //
+  // Requires: cancel() must happen after init() on a given timer
+  virtual void Cancel(grpc_exec_ctx* exec_ctx, TimerStorage* timer) = 0;
+
+  // Check for timers to be run, and run them.
+  // Return FIRED if timer callbacks were executed, CHECKED_AND_EMPTY if timers
+  // were checked but none were ready, and NOT_CHECKED if checks were skipped.
+  // If next is non-null, TRY to update *next with the next running timer IF
+  // that timer occurs before *next current value. *next is never guaranteed to
+  // be updated on any given execution; however, with high probability at least
+  // one thread in the system will see an update at any time slice.
+  enum class CheckResult { NOT_CHECKED, CHECKED_AND_EMPTY, FIRED };
+  virtual void CheckTimers(grpc_exec_ctx* exec_ctx, grpc_millis* next) = 0;
+
+  // Consume a kick received by TimerManager::Kick
+  virtual void ConsumeKick() = 0;
+};
+
+// Concrete Timer instance class. To allow this type to avoid memory allocation
+// even though the TimerEngine is abstract, we pre-size some storage memory and
+// just pass that through (via inlined functions) to the TimerEngine
+// implementation that's in use.
+class Timer {
+ public:
+  Timer(grpc_exec_ctx* exec_ctx, grpc_millis deadline,
+        grpc_closure* on_complete) {
+    TimerEngine::Get()->Init(exec_ctx, &storage_, deadline, on_complete);
```

Comment:
We need a way to create an unset timer since CQ alarms can't be set at construction time (since a different thread may pick up the cq event while the Timer object is still under construction).

Reply:
Why do we need to create the timer in the first place?

Reply:
Resolution: we can deal with this by having the alarm use a ManualConstructor on the timer that it wraps. That keeps the complexity in alarm rather than in timer.